### DISCUSSION TITLE
Build tools, React etc.

### DISCUSSION THREAD

**Discussion by xinaesthete** [Taylor-CCB-Group/MDV#44] [General]
One potential issue with starting to use React is that it can potentially be a bit of a hassle to configure - an issue I had in the past was starting with Create React App, which makes it easy to get going but can make things more complicated if you start wanting to do something it wasn't designed for... and if you `eject` to manage the configuration manually you're left with a bit of a convoluted mess and no way to go back. It feels like something that tries to hide things from you and gaslight you for wanting to deviate (or even just understand what it's doing in reasonably simple terms).

When I decided to experiment with React  (and `react-three-fiber`) for the ['virtual exhibition'](app.shemza.digital/r3f) thing I'm working on at the moment, it was pretty painless changing the existing Vite configuration to allow me to use React. I'm slightly loathe to suggest switching from Webpack; it'd probably be an equivalent small change to the configuration there... but past experience of changing Webpack configuration has been that I find it often hard to figure out how to get it right... thankfully, my recent addition of TypeScript went ok, hopefully React would be similar. That said, I have found Vite much faster and easier to work with, so I may also experiment with using that instead. I suspect the end result might be a bit nicer to work with, and it may even be quicker to make a new vite config than wrestle with Webpack, but I don't want to force unnecessary changes etc.

**Comment by martinSergeant** [comment]
Create React App is overkill (there are many memes about this e.g. https://9gag.com/gag/aZL5EZz) I just use 6 packages  (react, react-dom, babel/cli ,babel/core, babel/preset-react, babel-loader) and the following rule in the webpack config

```
{
   test: /\.jsx/,
   use:{
     loader: 'babel-loader',
     options: {
     	presets: [['@babel/preset-react',{"runtime": "automatic"}]]
     }
   }
}
```
I experimented using a react wrapper in the early days, but took it no further

```
import {getRandomDataStore} from "src/dev/MockDataStore.js";
import "src/charts/HistogramChart";
import BaseChart from "src/charts/BaseChart.js";
import {useRef,useEffect,useState} from "react";
import {render} from "react-dom";

const types  =  BaseChart.types;
function Chart(props){
  const div = useRef();
  const chart = useRef();
  useEffect(()=>{
    chart.current= new types[props.config.type].class(props.dataStore,div.current,props.config);
    return ()=>{
      chart.current.remove();
    }
  },[]);
 
  return (
    <div ref = {div}></div>
  )
}

const ds= getRandomDataStore(100000,{ranges:[[-20,50],[-20,50],[-20,50]]});
render(<Chart config={{type:"bar_chart",param:["x3"]}} dataStore={ds}/>,document.getElementById("root"));
```

If the switch over to vite is simple than we can try this. I don't seem to have problems with the speed of webpack - as recompiling usually takes < 1 sec as I guess it is smart and only recompiles code that has been changed.

Two things I couldn't do with webpack both concern webworkers - I could never get jest, webpack and webworkers to play nicely - I tried using a proxy  jsdom-worker but it didn't work.

A second is in production, it spits out random names  for the webworker js files. In the production app, I load the js files in the html  file with name.js?jsversion where jsversion is specified in a config to ensure stale cached  copies are not used by the user. I can't do the same for webworkers as they are called internally in the compiled code. The only work around I could fine was when I make changes to a webworker, I create it  with a new version number e.g.

`new Worker (new URL("./catColWorker.js?v=3",import.meta.url));`

This causes a new name to be given to the worker  js file when compiled and hence users will not get the  stale version

Perhaps vite may solve these problems

**Comment by xinaesthete** [comment]
Since I needed to somewhat revisit build config to get an [automated build of my branch](https://mdv-dev.netlify.app/) up and running I thought this might be a time to revisit Vite.

The only thing that needed any particular attention to get Vite config working was headers for Cross-Origin isolation needed for `SharedArrayBuffer`.

Currently, in my `pjt-dev`, it's just running my `obvios-example` since that has a simple config in terms of data etc. I'll expand this config a bit more, maybe adding a couple more HTML files for various mock-data examples - it'd be good to have something that tries to get reasonable coverage in terms of features. There's a new `index.html` which is currently the entry point for the vite config - that has

```
<script src="./src/modules/obvios_index.ts" type="module"></script>
```

I think I'd have needed to do something somewhat similar to get Webpack working in a way that would be able to be digested by Netlify / Vercel etc. I went with Netlify in this instance because it let me use my free account to deploy a public repo.

I might have a bit more of a look at worker-related issues, Vite may be able to clean things up a bit in that regard. I haven't tested this current build mechanism very extensively.

**Reply by xinaesthete** [reply]
I also added a very basic proof-of-concept React-based chart. I don't necessarily advocate pushing this too far or fast, but it did go reasonably smoothly - and it is *really* nice being able to edit code and have it update in realtime without losing state or having to reload a lot of data. I was pleasantly surprised to find that HMR happens without needing to change anything about the wider state-management architecture etc.

I didn't want to start bringing in any extra state-management libraries or anything here, or do anything that would require changes not isolated to a given component. This being added to the project shouldn't have any impact on anything that doesn't import it.

```jsx
let ID = 0;
function ReactTest({dataStore, parent}) {
  const [id] = useState(ID++);
  const [filterSize, setFilterSize] = useState(dataStore.filterSize);
  const [text, setText] = useState(parent.config.text);
  useEffect(() => {
    const listener = () => setFilterSize(dataStore.filterSize);
    dataStore.addListener(id, listener);
    parent.addListener("text", (type, data) => setText(data));
    return () => {
      dataStore.removeListener(id);
      parent.removeListener("text");
    }
  }, [dataStore]);
  return (
    <div>
      <h1>React Test</h1>
      <h2>filterSize: {filterSize}</h2>
      <p>{text}</p>
    </div>
  );
}

class ReactChart extends BaseChart {
  constructor(dataStore, div, config) {
    super(dataStore, div, config);
    createRoot(this.contentDiv).render(<ReactTest dataStore={dataStore} parent={this} />);
  }

  getSettings() {
    const c = this.config;
    const settings = super.getSettings();
    return settings.concat([
      {
        label:"Text",
        type:"textbox",
        current_value: c.text,
        func: x => {
          c.text = x;
          this._callListeners("text", x);
        }
      }
    ]);
  }
}

BaseChart.types["ReactChart"] = {
  "class": ReactChart,
  name: "React HelloWorld Chart",
  init: (config, dataSource, extraControls) => {
    config.text = "Hello World";    
  }
}
```

**Reply by xinaesthete** [reply]
OK, so I did another brief experiment with a React-based version of the `AnnotationDialog` I'd been working on a while ago - which I think could be useful once it has a better UI (not necessarily in React, but it's something I'd vaguely earmarked for this experiment). I am finding that HMR doesn't quite 'just work' in this instance as it did with my other `ReactChart` test. I think that should not be difficult to resolve.

I'm pretty sure the previous experiment had been able to work partly as a result of the way the chart is registered to `BaseChart.types` as a side-effect of importing the module, rather than being exported as a class/whatever and used in a way that the dependency graph is able to keep track of.

The result of this was that changes in the `ReactTest` function above (including things like changing which listeners were added, how that gets rendered etc) would be instantly reflected in the page, while changes to `ReactChart` class would not apply to existing instances of the chart, but would apply to new ones invoked after the change - which is about the best that could be hoped for, I think.

At the moment, my experimental `AnnotationDialogReact` is added to the chart manager menu like this (in my main JS index file):

```
import AnnotationDialogReact from "../charts/dialogs/AnnotationDialogReact";
chartManager.addMenuIcon(ds.name, "fas fa-tags", "Tag annotation React", () => { new AnnotationDialogReact(ds.dataStore, tagModel) });
```

When the system sees changes to `AnnotationDialogReact`, that propagates in a way that means the whole page has to reload. It seems like there's a benefit in this case to the 'impure' side-effects `BaseChart.type[name] = {"class": MyClass}` in side-stepping that dependency graph, using HMR in localised parts of the code without needing any overarching architectural features to support it, or potentially local bundler-specific code for accepting changes etc.

I really don't plan on spending much time on this stuff just at the moment, but I will make enough of a change to verify that. Hope it makes sense.
